File: app/repositories/memory_repository.py
"""Repository for the simplified user memory system."""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any

from app.core.exceptions import DatabaseError
from app.repositories.base import Repository


@dataclass
class UserMemory:
    """Represents a single memory for a user."""

    id: int
    user_id: int
    chat_id: int
    memory_text: str
    created_at: int
    updated_at: int


class MemoryRepository(Repository[UserMemory]):
    """Repository for managing user memories in the database."""

    async def init(self) -> None:
        """Initializes the repository by ensuring the table exists."""
        # This is a no-op if the schema is managed externally,
        # but good practice for repositories that might manage their own schema.
        pass

    async def find_by_id(self, id: Any) -> UserMemory | None:
        """Finds a memory by its primary key."""
        return await self.get_memory_by_id(id)

    async def save(self, entity: UserMemory) -> UserMemory:
        """Saves a memory entity. Not implemented, use add_memory."""
        raise NotImplementedError("Use add_memory for creating new memories.")

    async def delete(self, id: Any) -> bool:
        """Deletes a memory by its ID."""
        return await self.delete_memory(id)

    async def add_memory(
        self, user_id: int, chat_id: int, memory_text: str
    ) -> UserMemory:
        """
        Adds a new memory for a user.

        Args:
            user_id: The user's ID.
            chat_id: The chat's ID.
            memory_text: The text of the memory to add.

        Returns:
            The newly created UserMemory object.

        Raises:
            DatabaseError: If the memory already exists (UNIQUE constraint).

        Note:
            If the user has 15 memories, the oldest one is automatically deleted (FIFO).
        """
        now = int(datetime.now(UTC).timestamp())

        # First, check if the user is at the memory limit
        count_query = (
            "SELECT COUNT(*) FROM user_memories WHERE user_id = $1 AND chat_id = $2"
        )
        row = await self._fetch_one(count_query, (user_id, chat_id))
        if row and row[0] >= 15:
            # Auto-delete the oldest memory (FIFO) to make room
            delete_oldest_query = """
                DELETE FROM user_memories
                WHERE id = (
                    SELECT id FROM user_memories
                    WHERE user_id = $1 AND chat_id = $2
                    ORDER BY created_at ASC
                    LIMIT 1
                )
            """
            await self._execute(delete_oldest_query, (user_id, chat_id))

        # Insert the new memory with RETURNING clause to get the ID
        query = """
            INSERT INTO user_memories (user_id, chat_id, memory_text, created_at, updated_at)
            VALUES ($1, $2, $3, $4, $5)
            RETURNING id
        """
        try:
            row = await self._fetch_one(
                query, (user_id, chat_id, memory_text, now, now)
            )
            if not row or row["id"] is None:
                raise DatabaseError("Failed to get last row id.")
            return UserMemory(
                id=row["id"],
                user_id=user_id,
                chat_id=chat_id,
                memory_text=memory_text,
                created_at=now,
                updated_at=now,
            )
        except DatabaseError:
            raise
        except Exception as e:
            # The UNIQUE constraint on (user_id, chat_id, memory_text) might fail
            error_str = str(e).lower()
            if "unique constraint" in error_str or "duplicate key" in error_str:
                raise DatabaseError(
                    "This memory already exists for the user."
                ) from None
            raise DatabaseError(f"Failed to add memory: {e}") from e

    async def get_memories_for_user(
        self, user_id: int, chat_id: int, limit: int = 100, offset: int = 0
    ) -> list[UserMemory]:
        """
        Retrieves all memories for a given user in a specific chat.

        Args:
            user_id: The user's ID.
            chat_id: The chat's ID.
            limit: Maximum number of memories to return.
            offset: Number of memories to skip.

        Returns:
            A list of UserMemory objects.
        """
        query = "SELECT * FROM user_memories WHERE user_id = $1 AND chat_id = $2 ORDER BY created_at ASC LIMIT $3 OFFSET $4"
        rows = await self._fetch_all(query, (user_id, chat_id, limit, offset))
        return [UserMemory(**row) for row in rows]

    async def get_memory_by_id(self, memory_id: int) -> UserMemory | None:
        """
        Retrieves a single memory by its ID.

        Args:
            memory_id: The ID of the memory.

        Returns:
            A UserMemory object if found, otherwise None.
        """
        query = "SELECT * FROM user_memories WHERE id = $1"
        row = await self._fetch_one(query, (memory_id,))
        return UserMemory(**row) if row else None

    async def delete_memory(self, memory_id: int) -> bool:
        """
        Deletes a single memory by its ID.

        Args:
            memory_id: The ID of the memory to delete.

        Returns:
            True if a memory was deleted, False otherwise.
        """
        query = "DELETE FROM user_memories WHERE id = $1"
        result = await self._execute(query, (memory_id,))
        # PostgreSQL execute returns string like "DELETE 1" or "DELETE 0"
        # Extract the number to check if any rows were deleted
        match = re.search(r"DELETE (\d+)", result)
        if match:
            return int(match.group(1)) > 0
        return False

    async def delete_all_memories(self, user_id: int, chat_id: int) -> int:
        """
        Deletes all memories for a given user in a specific chat.

        Args:
            user_id: The user's ID.
            chat_id: The chat's ID.

        Returns:
            The number of memories deleted.
        """
        query = "DELETE FROM user_memories WHERE user_id = $1 AND chat_id = $2"
        result = await self._execute(query, (user_id, chat_id))
        # PostgreSQL execute returns string like "DELETE 5"
        # Extract the number to get count of deleted rows
        match = re.search(r"DELETE (\d+)", result)
        if match:
            return int(match.group(1))
        return 0

File: cpp/src/repositories/memory_repository.cpp
#include "gryag/repositories/memory_repository.hpp"

#include <SQLiteCpp/Statement.h>
#include <spdlog/spdlog.h>

#include <chrono>

namespace gryag::repositories {

MemoryRepository::MemoryRepository(std::shared_ptr<infrastructure::SQLiteConnection> connection)
    : connection_(std::move(connection)) {}

std::int64_t MemoryRepository::get_current_timestamp() {
    return std::chrono::duration_cast<std::chrono::seconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();
}

MemoryRepository::UserMemory MemoryRepository::row_to_memory(SQLite::Statement& stmt) {
    UserMemory memory;
    memory.id = stmt.getColumn("id").getInt();
    memory.user_id = stmt.getColumn("user_id").getInt64();
    memory.chat_id = stmt.getColumn("chat_id").getInt64();
    memory.memory_text = stmt.getColumn("memory_text").getString();
    memory.created_at = stmt.getColumn("created_at").getInt64();
    memory.updated_at = stmt.getColumn("updated_at").getInt64();
    return memory;
}

MemoryRepository::UserMemory MemoryRepository::add_memory(
    std::int64_t user_id,
    std::int64_t chat_id,
    const std::string& memory_text
) {
    const auto now = get_current_timestamp();

    try {
        // Check if user is at memory limit (15)
        SQLite::Statement count_stmt(
            connection_->db(),
            "SELECT COUNT(*) as count FROM user_memories WHERE user_id = ? AND chat_id = ?"
        );
        count_stmt.bind(1, user_id);
        count_stmt.bind(2, chat_id);

        if (count_stmt.executeStep()) {
            const int count = count_stmt.getColumn("count").getInt();
            if (count >= 15) {
                // Auto-delete oldest memory (FIFO) to make room
                SQLite::Statement delete_stmt(
                    connection_->db(),
                    R"(DELETE FROM user_memories
                       WHERE id = (
                           SELECT id FROM user_memories
                           WHERE user_id = ? AND chat_id = ?
                           ORDER BY created_at ASC
                           LIMIT 1
                       ))"
                );
                delete_stmt.bind(1, user_id);
                delete_stmt.bind(2, chat_id);
                delete_stmt.exec();

                spdlog::debug("Deleted oldest memory for user {} in chat {} (FIFO limit)", user_id, chat_id);
            }
        }

        // Insert the new memory
        SQLite::Statement insert_stmt(
            connection_->db(),
            R"(INSERT INTO user_memories (user_id, chat_id, memory_text, created_at, updated_at)
               VALUES (?, ?, ?, ?, ?))"
        );
        insert_stmt.bind(1, user_id);
        insert_stmt.bind(2, chat_id);
        insert_stmt.bind(3, memory_text);
        insert_stmt.bind(4, now);
        insert_stmt.bind(5, now);
        insert_stmt.exec();

        const int memory_id = static_cast<int>(connection_->db().getLastInsertRowid());

        spdlog::info("Added memory {} for user {} in chat {}", memory_id, user_id, chat_id);

        UserMemory memory;
        memory.id = memory_id;
        memory.user_id = user_id;
        memory.chat_id = chat_id;
        memory.memory_text = memory_text;
        memory.created_at = now;
        memory.updated_at = now;

        return memory;

    } catch (const SQLite::Exception& ex) {
        // Check for UNIQUE constraint failure
        if (std::string(ex.what()).find("UNIQUE constraint") != std::string::npos) {
            throw std::runtime_error("This memory already exists for the user");
        }
        spdlog::error("Failed to add memory: {}", ex.what());
        throw std::runtime_error(std::string("Failed to add memory: ") + ex.what());
    }
}

std::vector<MemoryRepository::UserMemory> MemoryRepository::get_memories_for_user(
    std::int64_t user_id,
    std::int64_t chat_id
) {
    std::vector<UserMemory> memories;

    try {
        SQLite::Statement stmt(
            connection_->db(),
            "SELECT * FROM user_memories WHERE user_id = ? AND chat_id = ? ORDER BY created_at ASC"
        );
        stmt.bind(1, user_id);
        stmt.bind(2, chat_id);

        while (stmt.executeStep()) {
            memories.push_back(row_to_memory(stmt));
        }
    } catch (const std::exception& ex) {
        spdlog::error("Error getting memories for user: {}", ex.what());
    }

    return memories;
}

std::optional<MemoryRepository::UserMemory> MemoryRepository::get_memory_by_id(int memory_id) {
    try {
        SQLite::Statement stmt(
            connection_->db(),
            "SELECT * FROM user_memories WHERE id = ?"
        );
        stmt.bind(1, memory_id);

        if (stmt.executeStep()) {
            return row_to_memory(stmt);
        }

        return std::nullopt;
    } catch (const std::exception& ex) {
        spdlog::error("Error getting memory by id: {}", ex.what());
        return std::nullopt;
    }
}

bool MemoryRepository::delete_memory(int memory_id) {
    try {
        SQLite::Statement stmt(
            connection_->db(),
            "DELETE FROM user_memories WHERE id = ?"
        );
        stmt.bind(1, memory_id);
        const int rows = stmt.exec();

        if (rows > 0) {
            spdlog::info("Deleted memory {}", memory_id);
            return true;
        }
        return false;
    } catch (const std::exception& ex) {
        spdlog::error("Error deleting memory: {}", ex.what());
        return false;
    }
}

int MemoryRepository::delete_all_memories(std::int64_t user_id, std::int64_t chat_id) {
    try {
        SQLite::Statement stmt(
            connection_->db(),
            "DELETE FROM user_memories WHERE user_id = ? AND chat_id = ?"
        );
        stmt.bind(1, user_id);
        stmt.bind(2, chat_id);
        const int rows = stmt.exec();

        spdlog::info("Deleted {} memories for user {} in chat {}", rows, user_id, chat_id);
        return rows;
    } catch (const std::exception& ex) {
        spdlog::error("Error deleting all memories: {}", ex.what());
        return 0;
    }
}

int MemoryRepository::get_memory_count(std::int64_t user_id, std::int64_t chat_id) {
    try {
        SQLite::Statement stmt(
            connection_->db(),
            "SELECT COUNT(*) as count FROM user_memories WHERE user_id = ? AND chat_id = ?"
        );
        stmt.bind(1, user_id);
        stmt.bind(2, chat_id);

        if (stmt.executeStep()) {
            return stmt.getColumn("count").getInt();
        }
        return 0;
    } catch (const std::exception& ex) {
        spdlog::error("Error getting memory count: {}", ex.what());
        return 0;
    }
}

bool MemoryRepository::update_memory(int memory_id, const std::string& new_text) {
    const auto now = get_current_timestamp();

    try {
        SQLite::Statement stmt(
            connection_->db(),
            "UPDATE user_memories SET memory_text = ?, updated_at = ? WHERE id = ?"
        );
        stmt.bind(1, new_text);
        stmt.bind(2, now);
        stmt.bind(3, memory_id);
        const int rows = stmt.exec();

        if (rows > 0) {
            spdlog::info("Updated memory {}", memory_id);
            return true;
        }
        return false;
    } catch (const std::exception& ex) {
        spdlog::error("Error updating memory: {}", ex.what());
        return false;
    }
}

}  // namespace gryag::repositories
